src/main/java/net/simpvp/Misc/DamageSource.java

```diff
@@ -0,0 +1,70 @@
+package net.simpvp.Misc;
+
+import org.bukkit.entity.*;
+import org.bukkit.event.EventHandler;
+import org.bukkit.event.Listener;
+import org.bukkit.event.entity.EntityDamageByEntityEvent;
+import org.bukkit.event.entity.EntityDamageEvent;
+import org.bukkit.event.entity.ExplosionPrimeEvent;
+import org.bukkit.scheduler.BukkitRunnable;
+import org.bukkit.util.Vector;
+
+/**
+ * Modifies damage source data in various ways
+ * to prevent it from being used to track players positions
+ *
+ * @author cats
+ */
+public class DamageSource implements Listener {
+
+    private final int maxDistanceSq;
+
+    public DamageSource(int maxDistance) {
+        this.maxDistanceSq = maxDistance * maxDistance;
+    }
+
+    /**
+     * This handler removes the TNT explosion source
+     * By doing this, we prevent the tnt knockback from an improper damage source
+     */
+    @EventHandler
+    public void onExplosionPrimed(ExplosionPrimeEvent event) {
+        if (event.getEntityType() == EntityType.PRIMED_TNT) {
+            final TNTPrimed tnt = (TNTPrimed) event.getEntity();
+
+            tnt.setSource(null);
+        }
+    }
+
+    /**
+     * Reset the velocity 1 tick after the player takes damage
+     *
+     * Possible todo: maybe make this run async to avoid tps lag playing a factor
+     * I don't think it does because bukkit scheduler, but who knows
+     */
+    @EventHandler
+    public void onDamage(EntityDamageByEntityEvent event) {
+        if (event.getCause() == EntityDamageEvent.DamageCause.PROJECTILE) {
+            // This SHOULD always be a successful cast
+            final Projectile projectile = (Projectile) event.getDamager();
+
+            if (projectile.getShooter() instanceof Player) {
+                final Player shooter = (Player) projectile.getShooter();
+
+
+                // I'm wincing but it shouldn't be too much math
+                if (projectile.getLocation().distanceSquared(shooter.getLocation()) > this.maxDistanceSq) {
+                    new BukkitRunnable() {
+                        @Override
+                        public void run() {
+                            event.getEntity().setVelocity(new Vector(0, event.getEntity().getVelocity().getY(), 0));
```

Comment:
Will this remove a player's existing velocity when they get hit by an arrow from far away?

Like if they're already moving for an unrelated reason, then this'll just instantly stop them?

I don't think it's important since it only triggers when the shooter is very far away, but you could get their velocity from the moment before they get hit by the projectile, and then reset their velocity to that at this point, so their velocity essentially just freezes for 1 tick.

Reply:
I thought about that, but I wanted to make it interrupt the current velocity like fire does. It seemed more reasonable than just having it deal no knockback. Also, the default distance is 300 blocks, which I think wouldn't be possible with most server's render distances.

Reply:
At one point I found myself poking through the spigot jar trying to recreate the standard arrow velocity calculation (when the shooter of the arrow is null), but it was just too messy, especially when I realized I would be making it function with the bukkit API.

Let me know if I can resolve this, or if you'd like some further development on this.